File: java/src/main/java/net/netheos/pcsapi/models/CDownloadRequest.java
/**
 * Copyright (c) 2014 Netheos (http://www.netheos.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.netheos.pcsapi.models;

import net.netheos.pcsapi.request.Headers;
import net.netheos.pcsapi.bytesio.ByteSink;
import net.netheos.pcsapi.bytesio.ProgressByteSink;
import net.netheos.pcsapi.bytesio.ProgressListener;

import org.apache.http.Header;
import org.apache.http.message.BasicHeader;

/**
 * Class that parametrizes a download request: path + bytes sink + (optional) byte range
 */
public class CDownloadRequest
{

    private final CPath path;
    private final ByteSink byteSink;
    private ByteRange byteRange;
    private ProgressListener progressListener;

    public CDownloadRequest( CPath path, ByteSink byteSink )
    {
        this.path = path;
        this.byteSink = byteSink;
        this.byteRange = null;
    }

    /**
     * Get the file path to download
     *
     * @return The file path
     */
    public CPath getPath()
    {
        return path;
    }

    /**
     * Get the HTTP headers to be used for download request.
     *
     * Default implementation only handles Range header.
     *
     * @return The headers
     */
    public Headers getHttpHeaders()
    {
        Headers headers = new Headers();

        if ( byteRange != null ) {
            StringBuilder rangeBuilder = new StringBuilder( "bytes=" );

            long start;
            if ( byteRange.offset >= 0 ) {
                rangeBuilder.append( byteRange.offset );
                start = byteRange.offset;
            } else {
                start = 1;
            }
            rangeBuilder.append( "-" );
            if ( byteRange.length > 0 ) {
                rangeBuilder.append( start + byteRange.length - 1 );
            }

            Header rangeHeader = new BasicHeader( "Range", rangeBuilder.toString() );
            headers.addHeader( rangeHeader );
        }
        return headers;
    }

    /**
     * Defines a range for partial content download. Note that second parameter is a length, not an offset (this differs
     * from http header Range header raw value).
     * <ul>
     * <li>(-1, -1) = default : download full blob.</li>
     * <li>(10, -1) = download, starting at offset 10</li>
     * <li>(-1, 10) = download last 10 bytes</li>
     * <li>(10, 100) = download 100 bytes, starting at offset 10.</li>
     * </ul>
     *
     * @param offset The start offset to download the file, or &lt;0 for downloading only end of file
     * @param length (never 0) The data length to download, or &lt;0 for downloading up to end of file
     * @return This download request
     */
    public CDownloadRequest setRange( long offset, long length )
    {
        if ( length == 0 ) {
            // Indicate we want to download 0 bytes ?! We ignore such specification
            length = -1;
        }
        if ( offset < 0 && length < 0 ) {
            byteRange = null;
        } else {
            byteRange = new ByteRange( offset, length );
        }
        return this;
    }

    /**
     * Defines an object that will be notified during download.
     *
     * @param pl the progress listener
     * @return this download request
     */
    public CDownloadRequest setProgressListener( ProgressListener pl )
    {
        this.progressListener = pl;
        return this;
    }

    /**
     * If no progress listener has been set, return the byte sink defined in constructor, otherwise decorate it.
     *
     * @return the byte sink to be used for download operations.
     */
    public ByteSink getByteSink()
    {
        ByteSink bs = byteSink;
        if ( progressListener != null ) {
            bs = new ProgressByteSink( bs, progressListener );
        }
        return bs;
    }

    private static class ByteRange
    {

        private final long offset;
        private final long length;

        public ByteRange( long offset, long length )
        {
            this.offset = offset;
            this.length = length;
        }

    }

}

File: cpp/libpcs_api/src/model/c_download_request.cc
/**
 * Copyright (c) 2014 Netheos (http://www.netheos.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <sstream>

#include "pcs_api/c_download_request.h"
#include "pcs_api/internal/progress_byte_sink.h"
#include "pcs_api/internal/logger.h"


namespace pcs_api {

CDownloadRequest::CDownloadRequest(CPath path,
                                   std::shared_ptr<ByteSink> p_byte_sink)
    : path_(path),
      p_byte_sink_(p_byte_sink),
      range_offset_(-1),
      range_length_(-1) {
}

CDownloadRequest& CDownloadRequest::set_progress_listener(
                                     std::shared_ptr<ProgressListener> p_pl) {
    p_listener_ = p_pl;
    return *this;
}

std::map<string_t, string_t> CDownloadRequest::GetHttpHeaders() const {
    std::map<string_t, string_t> headers;
    if (range_offset_ >= 0 || range_length_ > 0) {
        std::basic_ostringstream<char_t> range;
        range << PCS_API_STRING_T("bytes=");

        int64_t start;
        if (range_offset_ >= 0) {
            range << range_offset_;
            start = range_offset_;
        } else {
            start = 1;
        }
        range << PCS_API_STRING_T("-");
        if (range_length_ > 0) {
            range << (start + range_length_ - 1);
        }

        string_t header_value = range.str();
        LOG_TRACE << "Range: " << header_value;
        headers[PCS_API_STRING_T("Range")] = header_value;
    }
    return headers;
}

CDownloadRequest& CDownloadRequest::SetRange(int64_t offset, int64_t length) {
    if (length == 0) {
        // Indicate we want to download 0 bytes ?! We ignore such requests
        LOG_WARN << "Ignored range length setting of 0.";
        length = -1;
    }
    range_offset_ = offset;
    range_length_ = length;
    return *this;
}

std::shared_ptr<ByteSink> CDownloadRequest::GetByteSink() const {
    if (!p_listener_) {
        return p_byte_sink_;
    } else {
        return std::make_shared<detail::ProgressByteSink>(p_byte_sink_,
                                                          p_listener_);
    }
}

}  // namespace pcs_api
